zerver/lib/message.py

```diff
@@ -1633,6 +1635,19 @@ def wildcard_mention_allowed(sender: UserProfile, stream: Stream) -> bool:
     raise AssertionError("Invalid wildcard mention policy")
 
 
+def check_user_group_mention_allowed(sender: UserProfile, user_group_ids: List[int]) -> None:
+    user_groups = UserGroup.objects.filter(id__in=user_group_ids).select_related(
+        "can_mention_group"
+    )
+
+    for group in user_groups:
+        can_mention_group = group.can_mention_group
+        if not is_user_in_group(can_mention_group, sender, direct_member_only=False):
```

Comment:
This does a DB query in a loop. But am not sure how often we would see a large number of user groups mentioned in a single message. I tried replacing these with a bulk query but that seemed somewhat complicated due to recursive groups being involved especially when we would like to show the ID of the group which user cannot mention. I can try again if we really want to call `is_user_in_group` in a loop.

Reply:
Mmm, yeah -- this feels probably-fine to me, but I hear your worry.

Is `user_group_ids` ordered by the order of the mentions, so they at least get the message for the _first_ failing group, so we get some advantage over iterating over them?

Reply:
`mentions_user_group_ids` field of `MessageRenderingResult` object is a `set`, which I guess is unordered. Though this function receives a list as parameter which is obtained by applyig `list()` on the set and I did some print debugging and using `list()` probably always results in a increasing order of group IDs.

Reply:
That feels like an implementation detail that we don't want to rely on.  Though "sorted by internal ID" isn't going to be differentiable by users from "sorted randomly," so I don't think there'd be much utility to sorting, here, other than generating a consistent error for the same set of groups.

If we decide we care about order, we could switch the `mentions_user_group_ids ` `set` to a `dict`, which would preserve insertion order.

Reply:
I don't think there is any real advantage of having `user_group_ids` in the same order as in the message. We cannot be sure at what position the user will mention the group they are not allowed to mention.

Reply:
Hmm. I was going to say that this seems OK because you're limited in how much resources you can consume by the number of user groups on the server... but that could end up being a few hundred. I think we're probably OK for merging this, but cleaning this up should be a high priority follow-up.

I think the best implementation for this check would be to have a function similar in structure to `get_recursive_subgroups_for_groups` that can do the query "Is this user a member of all of these groups". One theory would be to define such a function in `zerver/lib/user_groups.py` that uses this inefficient implementation to let us integrate this PR, and then start a #backend thread about how to write that function efficiently.